### ISSUE TITLE
[RFC] Make all data collecting plugins external

### ISSUE THREAD

**Issue by vlvkobal** [netdata/netdata#6771]
##### Question summary
It might be worth rewriting all internal plugins and make them external. It will make Netdata invulnerable to segmentation faults and deadlocks that might occur on the data collection side. On the other hand, it will introduce communication overhead between the plugin and the core Netdata.

Related: #6741 

##### OS / Environment
All supported operating systems

##### Component Name
internal plugins

**Comment by Ferroin** [comment]
So, worded another way, switch to a micro-kernel type architecture?  I think this is probably a good idea for reasons other than just the reliability aspect, namely that it will let users set specific privilege levels for individual plugins.

For the messaging side of things, I'm not sure that will be as much of an issue as you might think, if things are de-aggregated correctly and we are willing to have a slight delay between data being collected/saved and being displayed in the web UI.

**Comment by mfundul** [comment]
On the other hand increasing software overhead is not desirable when we're interested in very small devices and millisecond data collection intervals.

**Comment by Ferroin** [comment]
> On the other hand increasing software overhead is not desirable when we're interested in very small devices and millisecond data collection intervals.

Looking at that from the other direction though, if we split each plugin to it's own executable/process/whatever, we'll probably see a significant improvement in the memory footprint for the common case of not actually using most of the plugins on any given system.  It would also make it much easier for users to 'customize' their Netdata installs to only include what they actually need, thus simplifying the process of reducing the required space in persistent storage as well.

Also, as mentioned above, if we decouple the reporting intervals (that is, the intervals at which the web UI gets updated, and data gets sent to back-ends and master nodes) from the collection intervals, we can probably keep the communications overhead pretty low overall.

**Comment by Saruspete** [comment]
Agreed with @mfundul : when working on environment tight on resources (IoT) or sensible to latency (HPC, HFT) we'll select carefully the metrics we need, to avoid resource consumption for administrative tools and keep the maximum for business stuff.

While this is not the use-case for the most common types, I would need to be able to select it as "one single efficient process" rather than "many processes communicating each other".
You can default to "one process per plugin", but many will need to switch back to the "one process, many threads" mode.

The most ugly issue I see is with memory starvation and the sacrifice (OOM) of plugins. How will a killed plugin be reported ? Restarted ?

**Comment by Ferroin** [comment]
> You can default to "one process per plugin", but many will need to switch back to the "one process, many threads" mode.

I think it really depends on how, exactly, we split things out.  Stuff like cpuidle and cpufreq that aren't likely to be used by everybody should definitely be isolated, but there's no real reason to have the 'core' stats (CPU/RAM/Swap/Disk) split out individually (so, at most, one process that collects all of that, possibly just keep it in the core).

> The most ugly issue I see is with memory starvation and the sacrifice (OOM) of plugins. How will a killed plugin be reported ? Restarted ?

Presumably the same way that pretty much all software handles this, have a 'master' process that watches all the others and restarts them if they die.  On UNIX systems, you just need to keep track of what child processes you've started and watch for and handle `SIGCHILD` properly.  On Windows, I'm not certain, but looking into how Apache or nginx do things there may be worth some time.
